### PR TITLE
fix: integrate with aegir types cmd

### PR DESCRIPTION
**Pull request by hugomrdias**
This PR does the following: (review might be easier by commit)
- chore: update readme to look like the other multiformats repos
- fix: update package.json and tsconfig
- fix: transform multiaddr into a class

   BREAKING CHANGE: index.js default export no longer is both callable and
   constructor. encapsulate and decapsulate only accept Multiaddr instances as
   input.

- fix: remove class-is and fix types in index.js
- fix: add some types to protocols
- fix: add some types to convert

   BREAKING CHANGE: Convert.toString now always returns a string

- fix: add some types to codec.js
- fix: fix resolver exports
- fix: remove old docs stuff
- feat: use github ci

The breaking changes should be theoretically based on the current documentation and usage but we never know if people are doing something not documented. 

ping this PR https://github.com/multiformats/js-multiaddr/pull/168

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -4,556 +4,608 @@ const codec = require('./codec')
...
-Multiaddr.prototype.encapsulate = function encapsulate (addr) {
-  addr = Multiaddr(addr)
-  return Multiaddr(this.toString() + addr.toString())
-}
+  /**
+   * Returns a tuple of string/number parts
+   * - tuples[][0] = code of protocol
+   * - tuples[][1] = contents of address
+   *
+   * @returns {[number, string?][]} tuples
```
Comment (vasco-santos) [inline src/index.js:207]:
```suggestion
   * @returns {[number, string|number?][]} tuples
```

File: src/index.js [hunk 2]
```diff
@@ -4,556 +4,608 @@ const codec = require('./codec')
...
+   * @param {Multiaddr} [addr] - Defaults to using `this` instance
+   * @returns {boolean} isThinWaistAddress
+   * @example
+   * ```js
+   * const mh1 = Multiaddr('/ip4/127.0.0.1/tcp/4001')
+   * // <Multiaddr 047f000001060fa1 - /ip4/127.0.0.1/tcp/4001>
+   * const mh2 = Multiaddr('/ip4/192.168.2.1/tcp/5001')
+   * // <Multiaddr 04c0a80201061389 - /ip4/192.168.2.1/tcp/5001>
+   * const mh3 = mh1.encapsulate(mh2)
+   * // <Multiaddr 047f000001060fa104c0a80201061389 - /ip4/127.0.0.1/tcp/4001/ip4/192.168.2.1/tcp/5001>
```
Comment (vasco-santos) [inline src/index.js:482]: Per the TODO above: you can for example add a dnsaddr multiaddr or a webrtc-star multiaddr here

File: src/index.js [hunk 3]
```diff
@@ -4,556 +4,608 @@ const codec = require('./codec')
...
-  if (protos[0].code !== 4 && protos[0].code !== 41) {
-    return false
+  /**
+   * Returns an array of multiaddrs, by resolving the multiaddr that is a name
+   *
+   * @async
+   * @param {Multiaddr} addr
+   * @returns {Promise<Multiaddr[]>}
+   */
+  static resolve (addr) {
```
Comment (vasco-santos) [inline src/index.js:560]: This can be removed right? I did not see this when I implemented the `multiaddr.resolve()` and I think we can just leverage that.

File: src/protocols-table.js [hunk 4]
```diff
@@ -1,13 +1,20 @@
 'use strict'
+/** @typedef {import("./types").Protocol} Protocol */
 
+/**
+ * Protocols
+ *
+ * @param {number | string | string} proto
```
Comment (vasco-santos) [inline src/protocols-table.js:7]:
```suggestion
 * @param {number | string} proto
```

File: src/resolvers/dns.js [hunk 5]
```diff
@@ -3,7 +3,7 @@
 let dns
 
 try {
-  dns = require('dns').promises
+  dns = require('dns').Resolver
```
Comment (vasco-santos) [inline src/resolvers/dns.js:6]: Why this change? This should be using the promises API https://nodejs.org/dist/latest-v14.x/docs/api/dns.html#dns_dns_promises_api

File: src/resolvers/index.js [hunk 6]
```diff
@@ -1,25 +1,24 @@
...
 const protocols = require('../protocols-table')
 
 const { code: dnsaddrCode } = protocols('dnsaddr')
 
+// TODO `addr` type needs https://github.com/microsoft/TypeScript/issues/41672
 /**
  * Resolver for dnsaddr addresses.
  *
- * @param {Multiaddr} addr
+ * @param {any} addr
```
Comment (vasco-santos) [inline src/resolvers/index.js:11]: Why not a multiaddr?

**Review by vasco-santos — CHANGES_REQUESTED** [review]

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+name: Bug report
+about: Create a report to help us improve
+title: ''
+labels: ''
+assignees: ''
+
+---
+
+**Describe the bug**
+A clear and concise description of what the bug is.
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply .github/ISSUE_TEMPLATE/bug_report.md:11]: Probably, i just copy/pasted our base templates. i dont know who created these but are used in other repos. 
Maybe we can just rethink these in aegir and make a cmd to normalise everything ?

File: package.json [hunk 8]
```diff
@@ -4,13 +4,21 @@
...
+        "dist/src/*/index"
+      ]
+    }
+  },
   "scripts": {
     "lint": "aegir lint",
-    "test": "npm run test:node && npm run test:browser",
-    "test:node": "aegir test --ts -t node",
+    "test": "aegir test",
+    "test:node": "aegir test -t node",
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply package.json:19]: yeah i known i read through those.
aegir doesnt support mixed repos like the current master, this PR fixes all of that.

File: src/codec.js [hunk 9]
```diff
@@ -137,7 +144,14 @@ function sizeForAddr (p, addr) {
 }
 
 // Uint8Array -> [[int code, Uint8Array ]... ]
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/codec.js:146]: yeah i didnt add types for everything, this PR is focused in the public API been well typed. 
maybe we can do a follow up PR, this one is already big enough, plus the public api already protects type errors for this function.

File: src/index.js [hunk 10]
```diff
@@ -4,556 +4,608 @@ const codec = require('./codec')
...
-    /**
-     * @type {Uint8Array} - The raw bytes representing this multiaddress
-     */
-    this.bytes = codec.fromBytes(addr)
-  } else if (typeof addr === 'string' || addr instanceof String) {
-    if (addr.length > 0 && addr.charAt(0) !== '/') {
-      throw new Error(`multiaddr "${addr}" must start with a "/"`)
+    if (addr instanceof Uint8Array) {
+      /**
+       * @type {Uint8Array} - The raw bytes representing this multiaddress
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/index.js:39]: the type is already inferred from that method, but you still need this comment to get a proper documentation.
TS forces us to be really verbose with these things ... the public api needs to always have all the jsdoc and descriptions in place.

File: src/index.js (hunk 1 above)
Comment (hugomrdias) [reply src/index.js:207]: this is one of the "breaking changes",now at least the value is always a string.

`tuplesToStringTuples` was documented as `[string,string][]` but the tests expect `[number, string|number][]` so i made it at least `[number, string][]` because index 2 comes from `convert.toString` (port was the only thing not convert to string).

File: src/index.js (hunk 3 above)
Comment (hugomrdias) [reply src/index.js]: yes

File: src/resolvers/index.js (hunk 6 above)
Comment (hugomrdias) [reply src/resolvers/index.js:11]: theres a TODO above explaining we can't get the Multiaddr type because of this issue https://github.com/microsoft/TypeScript/issues/41672

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (vasco-santos) [reply .github/ISSUE_TEMPLATE/bug_report.md:11]: That would be good. Aegir could have a base template and the repos could overwrite them

File: src/index.js (hunk 1 above)
Comment (vasco-santos) [reply src/index.js:207]: That sounds good to me. I noticed this because of the example below. We can do the breaking change, just need to update the example below to have the port as a string

File: src/resolvers/index.js (hunk 6 above)
Comment (vasco-santos) [reply src/resolvers/index.js:11]: Ah right! This is not nice as we will need to update this again after, I might wait for this to be solved before getting libp2p PR updating this merged in.
But we can move forward with this module here.

File: src/index.js (hunk 2 above)
Comment (vasco-santos) [reply src/index.js:482]: Can you add here for example: `/ip4/127.0.0.1/tcp/2000/wss/p2p-webrtc-star/p2p/QmcgpsyWgH8Y8ajJz1Cu72KnS5uo2Aa2LpzU7kinSooo2a` 
You can also remove the TODO above

File: src/resolvers/index.js (hunk 6 above)
Comment (hugomrdias) [reply src/resolvers/index.js:11]: the only way to solve this now, is to do a bigger breaking change and default export the Multiaddr class directly. this would remove the ability to `multiaddr('/ip..')` and for `new Multiaddr('')`

should we do this ?

**Review by vasco-santos — COMMENTED** [review]
One minor detail left, and this should be good to go

File: src/resolvers/index.js (hunk 6 above)
Comment (hugomrdias) [reply src/resolvers/index.js:11]: i can add a static method to the Multiaddr class called `multiaddr` and users would only need to change from `const multiaddr = require('multiaddr)` to `const { multiaddr } = require('multiaddr)`

Comment (vasco-santos) [reply src/resolvers/index.js:11]: I would like to have the constructor exported by default to be honest. But, I would wait on @jacobheun input before moving that way as this has a impact on a large number of modules.

Comment (hugomrdias) [reply src/resolvers/index.js:11]: yeah having the constructor exported directly makes a bunch of things easier and documentation looks a lot better.

File: src/index.js (hunk 2 above)
Comment (hugomrdias) [reply src/index.js:482]: can i please suggest a change so i dont make mistakes ?

File: src/resolvers/index.js (hunk 6 above)
Comment (jacobheun) [reply src/resolvers/index.js:11]: I'm in favor of exporting the constructor, we should make sure we have clear migration docs for people. We should also add a static method for creation, this would make it easy to hot swap the two and allow us to be more picky about the constructor args. Since this is in the multiformats org it might be good to be consistent with the cid update effort, https://github.com/multiformats/js-multiformats/pull/23, and call it `asMultiaddr`

**Comment by hugomrdias** [comment]
@jacobheun and @vasco-santos can you please review, added the migration docs and some more small tweaks as for the `asMultiaddr` i would like to do that in another PR.

File: src/index.js [hunk 11]
```diff
@@ -4,556 +4,582 @@ const codec = require('./codec')
...
+    return value instanceof Multiaddr || Boolean(value && value[symbol])
   }
 
-  /*
-   * Needs more consideration from spec design:
-   *   - what to return
-   *   - how to achieve it in the browser?
+  /**
+   * Static factory method
+   *
```
Comment (Gozala) [inline src/index.js:572]: I think it would make sense to deprecate it to encourage transition.

```suggestion
   *
   * @deprecated
```

**Review by Gozala — CHANGES_REQUESTED** [review]
I made some small suggestions and some not so small. Given that this is breaking change, I think it would be good to use this opportunity to do more changes that otherwise would be a lot harder to justify:

1. Constructor takes address in arbitrary representation, tries to figure out what is it and then turns it into `Multiaddr` if it can. I think a factory functions (static method) representation kind is lot more clear both for user and implementer (as illustrated inline) and avoids unnecessary work when types are known. If it is not unreasonable I would really love to use this opportunity to do such a change. We can still have generic `Multiaddr.from` when representation is unknown.
2. `Multiaddr` has large number of utility methods, having equivalent static methods would be really nice to have because:
  - They come handy in places like `addrs.map(Multiaddr.toOptions)`
  - Moving `multiaddrs` across realms / threads discards prototype chains. Which means static methods continue to work, but not instance methods. Which in turn means encode / decode pipeline.

I realize above two are big asks & only bringing them so we can consider / discuss given the opportunity presented by breaking change.
